Added Inventory Page

justin-zhu1018 opened the pull request:
**Description**
A page to keep track of inventory for items that will be sold at the SCE shop. Basic backend CRUD functions, search bar to look items up, table to keep track of each item. Frontend does scale. Also created/edited the backend and frontend tests. 

**Testing Features**
1. To **view the Inventory Page**, log in to an admin account and click the "Inventory Page" tab under the Admin Dashboard. It should look something like this, but empty.
![image](https://user-images.githubusercontent.com/25803515/103488513-ffc41700-4dc1-11eb-9dcd-8ca80d7a0331.png)

2. To **add an item**, click the "+" icon to the right of the search bar. A modal form should pop up. Fill it with the information necessary. Price and Stock must be numerical values. Item Name should be unique (not case-sensitive). The only non-required input is the Picture, it can be left blank for a default "SCE Item" picture as a filler.
![image](https://user-images.githubusercontent.com/25803515/103488532-3732c380-4dc2-11eb-9606-c77bda510f8f.png)

3. To **edit an item**, click the pencil+paper icon on the furthest right column of your item in the table. A modal form should pop up with your existing information. Edit it as you please and click the "Confirm" button. 
![image](https://user-images.githubusercontent.com/25803515/103488610-e1aae680-4dc2-11eb-93a0-5f27f5e5c452.png)
![image](https://user-images.githubusercontent.com/25803515/103488618-e96a8b00-4dc2-11eb-8a14-c943b54a3c52.png)
![image](https://user-images.githubusercontent.com/25803515/103488634-0dc66780-4dc3-11eb-8361-92ea5fd2944d.png)
![image](https://user-images.githubusercontent.com/25803515/103488651-2898dc00-4dc3-11eb-8e11-2a06cab59345.png)

4. To **delete an item**, click the pencil+paper icon on the furthest right column of your item in the table. A modal form should pop up with your existing information. At the bottom of the modal, there should be a "Delete" button (reference screenshots in testcase #3). 

5. To **search for an item**, literally just start typing in the search bar. Keep in mind that it filters by item name.
-----------------------------
If there are any issues, please let me know 😊

IsitaB commented on api/main_endpoints/routes/InventoryItem.js:
```diff
@@ -20,16 +20,15 @@ router.get('/getItems', (req, res) => {
   InventoryItem.find(category).then(items => res.status(OK).send(items));
 });
 
+// TO-DO Find by _id
```
delet comment

aashrav reviewed (changes requested):
1. Categories should be drop down(electronics, snacks)
2. Description shouldn't be required (If there is no description the cell should say "None")
3. Indicate which fields are required which an asterisks next to the field

IsitaB commented on src/Pages/Inventory/InventoryPage.js:
```diff
@@ -0,0 +1,283 @@
+      picture: '',
+      inventoryItems:[],
+      allInventoryItems:[],
+      invalidPicture: 'https://user-images.githubusercontent.com/' +
```
lets get a constants.js file for /src and add this to it

IsitaB reviewed (changes requested):
prettiest modal on the website

could we round the price to two decimal places

form validation for modals: confirm button should always be clickable; on click render errors if required fields missing; row is as tall as the description (is this what we want?); add asterisks on required fields

add confirmation modal for delete button to be safe

item creation modal's create button is a little faded; consider making it darker

IsitaB reviewed (commented):
take package-lock out of your PR if you haven't added any dependencies

charlotte-zhuang commented on src/Pages/Inventory/InventoryPage.js:
```diff
@@ -0,0 +1,283 @@
+  }
+
+  // Function for add item form/modal
+  handleAddItem = async () => {
```
Trim input

charlotte-zhuang replied:
and validate input too

charlotte-zhuang commented on src/Pages/Inventory/InventoryPage.js:
```diff
@@ -0,0 +1,283 @@
+  }
+
+  // Function for edit item form/modal in table
+  handleEditItem = async (itemId) => {
```
Validate input

charlotte-zhuang commented on src/Pages/Inventory/EditItem/EditItemForm.js:
```diff
@@ -0,0 +1,91 @@
+          </FormText>
+        </Label>
+        <Input
+          invalid={isNaN(props.price)}
```
Also check for negative numbers

charlotte-zhuang reviewed (changes requested):
These comments might be outdated.

justin-zhu1018 commented:
All change requested so far should be covered in this commit. Let me know if I missed anything. 👍

**TO-DO:**
1. Category -> Dropdown
2. Empty Description -> "N/A"
3. Asterisks next to required fields in modals
4. Price Rounding
5. Form validation + alerts
6. Confirm button no longer disabled
7. Confirmation modal on delete button
8. Button color changes
9. A bunch of trimming/validation stuff
10. Old comments removed from code

**Updated Pictures for Reference:**
![image](https://user-images.githubusercontent.com/25803515/104690390-369b0680-56b9-11eb-916b-dc08f82baea4.png)
![image](https://user-images.githubusercontent.com/25803515/104690281-fe93c380-56b8-11eb-9f6f-124669ec09eb.png)
![image](https://user-images.githubusercontent.com/25803515/104690434-487ca980-56b9-11eb-93e4-d98a9ff6d116.png)
![image](https://user-images.githubusercontent.com/25803515/104690482-5cc0a680-56b9-11eb-867f-1928885782ea.png)
![image](https://user-images.githubusercontent.com/25803515/104690532-7366fd80-56b9-11eb-91c8-5eb7b03925c6.png)

aashrav reviewed (changes requested):
1.  In the form when you input a negative number for stock/price it complains, but it still lets you add the item with those invalid fields
2. When you do not fill out all of the required fields it throws an alert at the top of the modal. It would be nicer if the message showed above the Confirm & Delete button. The reason is because when I try to add an invalid item I have no idea whats wrong until I scroll to the top of the modal

justin-zhu1018 commented:
Implemented Aash's requested changes, all tests passed on my end, rebase and lint errors resolved.

Let me know if there are any other issues 👍

justin-zhu1018 commented:
Recent force-push for rebase

charlotte-zhuang reviewed (commented):
I think it looks good, except on mobile (the inventory table rows cut off content and it becomes unusable).

IsitaB commented:
> I think it looks good, except on mobile (the inventory table rows cut off content and it becomes unusable).

Superior UI. Excellent work

Resizing after a certain point cuts off column names. Please fix that

justin-zhu1018 commented:
Aash and I fixed the responsive issue by removing some of the less important columns (Image, description) starting at a certain screen size. There shouldn't be any more issues with that, but let me know if there is anything else. 👍

charlotte-zhuang reviewed (commented):
It's better, but text is still overflowing on mobile. Is it possible to turn the rows into flex-boxes and allow each item to use more rows when it runs out of space? Or maybe just turn the content vertical on mobile? The column headers would also have to change in that case.

Also, notice that the placeholder text is cut off in the search box.

<img width="756" alt="Screen Shot 2021-03-22 at 10 04 53 PM" src="https://user-images.githubusercontent.com/61573792/112081903-cc8c5b00-8b5a-11eb-9d17-9a4e25748a2b.png">

justin-zhu1018 commented:
Improved mobile view where each row has its own block rather than having it all squished in a table row.

IsitaB reviewed (commented):
<img width="985" alt="Screen Shot 2021-04-05 at 6 49 42 PM" src="https://user-images.githubusercontent.com/50391828/113647600-fd32c100-963f-11eb-8ea6-3238144ef469.png">
the div height is determined by the image. could we do uniform heights?

charlotte-zhuang reviewed (approved):
Best page on SCE?

Thai probably doesn't want you to commit `package-lock.json`

There's also a dependency conflict in `package.json`, but it's beyond the scope of this PR.
fixed: `"@tinymce/tinymce-react": "~3.8.4",`

justin-zhu1018 commented:
Please let me know if I missed anything. Thanks 👍

**QoL Stuff**
1. SVG buttons (add item, edit item) had a weird outline around them after clicking, so I removed those for a cleaner look.
2. Description Inputs in modals were able to be resized to an unusable height. Added min height styling.

**Table Responsiveness Fixes**
1. Removed outer container div that originally contained search and table component (margin: auto on outer container was making huge left/right margins at a certain screen size). Switched to 5% margins on left/right side for search and table containers individually.
2. Uniform heights on rows.  Fixed to a 100px uniform height. Overflow text/images are scrollable. Slight visual issue is that with scrollable, the images do get hidden horizontally at around 910px screen size. 

**Charlotte's dependency conflicts**
1. Not too sure what to do about the package-lock.json, but lmk what I need to do.
2. Seems like dev branch's tinymce is up to date now? `"@tinymce/tinymce-react": "^3.10.2"`.

aashrav reviewed (approved):
Looks good

charlotte-zhuang commented on src/Pages/Inventory/InventoryPage.css:
```diff
@@ -0,0 +1,127 @@
+.table-contents {
+  width: 100%;
+  height: 100px;
+  overflow: scroll;
```
This is causing issues on Firefox. Is it possible to truncate instead?

<img width="1214" alt="Screen Shot 2021-05-08 at 2 24 07 PM" src="https://user-images.githubusercontent.com/61573792/117549618-6cb61a80-b009-11eb-9414-6b5aaf3725df.png">

charlotte-zhuang replied:
[There are options to hide scrollbars](https://www.w3schools.com/howto/howto_css_hide_scrollbars.asp), but I don't think we really want scrolling at all.

charlotte-zhuang reviewed (changes requested):
See comment about `overflow: scroll`. I think valid alternatives are to hide overflow or go back to variable sized table rows. Ideally the image will be resized to fit x or y, centered, and overflow hidden.

We should do `git checkout dev package-lock.json` so that the file doesn't change, as per Thai's requests.

About `@tinymce/tinymce-react`: the correct version is `~3.8.4`, but don't worry about that.

justin-zhu1018 commented:
### Description
Pretty straightforward changes. Add character limit to description input in modal. Adjust responsiveness of table to remove description column earlier (at 1100px rather than 768px). Screenshots below.

### Screenshots
<img width="638" alt="CharLimit1" src="https://user-images.githubusercontent.com/25803515/134241558-4251b8c7-e913-48ad-a694-488c127bdf0e.png">
<img width="640" alt="CharLimit2" src="https://user-images.githubusercontent.com/25803515/134241569-8be40be4-b504-4757-85d0-3f3496387a33.png">
<img width="958" alt="Table1" src="https://user-images.githubusercontent.com/25803515/134241580-eb7b0d95-a05a-499c-a949-1c2701758c82.png">
<img width="545" alt="Table2" src="https://user-images.githubusercontent.com/25803515/134241584-a0679fd9-247c-4653-b3c4-463b1d0d94c7.png">
<img width="379" alt="Table3" src="https://user-images.githubusercontent.com/25803515/134241595-9eb2c8e1-55c0-4027-8569-de3ee3d0e4d7.png">
<img width="377" alt="Table4" src="https://user-images.githubusercontent.com/25803515/134241611-f51e8c31-ca54-4bca-9909-38d6cd5afc47.png">

andrewtedi commented on src/Pages/Inventory/AddItem/AddItemForm.js:
```diff
@@ -0,0 +1,110 @@
+          value={props.category}
+          onChange={(e) => props.updateItemCategory(e.target.value)}
+        >
+          <option value="">Choose...</option>
```
Add some more categories here? (Tools, Other)

andrewtedi reviewed (approved):
Looks really good, only add suggested changes if you really want to.
